graph/src/components/metrics/stopwatch.rs

```diff
@@ -0,0 +1,169 @@
+use crate::prelude::*;
+use std::collections::HashMap;
+use std::sync::{atomic::AtomicBool, atomic::Ordering, Mutex};
+use std::time::Instant;
+
+/// This is a "section guard", that closes the section on drop.
+pub struct Section {
+    id: String,
+    stopwatch: StopwatchMetrics,
+}
+
+impl Section {
+    /// A more readable `drop`.
+    pub fn end(self) {}
+}
+
+impl Drop for Section {
+    fn drop(&mut self) {
+        self.stopwatch
+            .end_section(std::mem::replace(&mut self.id, String::new()))
+    }
+}
+
+/// Usage example:
+/// ```ignore
+/// // Start counting time for the "main_section".
+/// let _main_section = stopwatch.start_section("main_section");
+/// // do stuff...
+/// // Pause timer for "main_section", start for "child_section".
+/// let chid_section = stopwatch.start_section("child_section");
+/// // do stuff...
+/// // Register time spent in "child_section", implicitly going back to "main_section".
+/// section.end();
+/// // do stuff...
+/// // At the end of the scope `_main_section` is dropped, which is equivalent to calling
+/// // `_main_section.end()`.
+#[derive(Clone)]
+pub struct StopwatchMetrics {
+    disabled: Arc<AtomicBool>,
+    inner: Arc<Mutex<StopwatchInner>>,
+}
+
+impl StopwatchMetrics {
+    pub fn new(
+        logger: Logger,
+        subgraph_id: SubgraphDeploymentId,
+        registry: Arc<dyn MetricsRegistry>,
+    ) -> Self {
+        let mut inner = StopwatchInner {
+            total_counter: *registry
+                .new_counter(
+                    format!("{}_total_secs", subgraph_id),
+                    format!("total time spent syncing"),
+                    HashMap::new(),
+                )
+                .expect("failed to register total_secs prometheus counter"),
+            logger,
+            subgraph_id,
+            registry,
+            counters: HashMap::new(),
+            section_stack: Vec::new(),
+            timer: Instant::now(),
+        };
+
+        // Start a base section so that all time is accounted for.
+        inner.start_section("unknown".to_owned());
+
+        StopwatchMetrics {
+            disabled: Arc::new(AtomicBool::new(false)),
+            inner: Arc::new(Mutex::new(inner)),
+        }
+    }
+
+    pub fn start_section(&self, id: &str) -> Section {
+        let id = id.to_owned();
+        if !self.disabled.load(Ordering::SeqCst) {
+            self.inner.lock().unwrap().start_section(id.clone())
+        }
+        Section {
+            id,
+            stopwatch: self.clone(),
+        }
```

Comment:
I don't understand this part yet: what does it mean to start a section on `inner` if the section is disabled and what does it mean to clone it? How does this make child sections work?

Deserves some explanatory comments I feel 😄

Reply:
I've added comments, if the stopwatch is disabled the section won't be started and nothing will happen when the returned value is dropped.